apps/synapse-portal-framework/src/shared-config/DoiRedirectRoute.tsx

```diff
@@ -0,0 +1,55 @@
+import { RouteObject, Navigate, useSearchParams } from 'react-router'
+
+export type DoiRedirectConfig<TResourceType extends string> = (
+  resourceType: TResourceType,
+  resourceId: string,
+) => string
+
+/**
+ * Component to redirect the Portal DOI ID to the resource in the portal.
+ * Portal DOI IDs are of the form "RESOURCETYPE.RESOURCEID". For example, `STUDY.syn123`.
+ *
+ * If no id is provided in the URLSearchParams, the user is redirected to the home page.
+ * @param props contains the redirectConfig function that maps the resource type/ID to the URL of the resource in the portal
+ */
+function DoiRedirectComponent<TResourceType extends string>(props: {
+  redirectConfig: DoiRedirectConfig<TResourceType>
+}) {
+  const { redirectConfig } = props
+  const [searchParams] = useSearchParams()
+  const doiId = searchParams.get('id')
+
+  if (!doiId) {
+    console.error(
+      'Could not redirect to DOI: no id provided in the URLSearchParams.',
+    )
+    return <Navigate to={'/'} replace={true} />
+  }
+
+  // Only split on the first `.` -- the ID may contain `.` itself
+  const parts = doiId.split('.', 2)
```

Comment:
Are we OK with this ID format? I think with the new way synapse handles DOIs, this could be changed in the future (I think the Portal DOI URI is an opaque string), but it might be a huge pain to migrate all of them.

Reply:
If we need to add another piece of information to this ID, how easy would it be?
I think on one extreme, we could use json.  On the other, a simple dot (or colon) separator for the tokens.
Should we lean into the URL encoded nature of this ID?  That is, what if the ID was something like "type=STUDY&id=syn123"?

Reply:
Another consideration is that our chosen format has to consistently serialize to equal strings. So if we use JSON/URLSearchParams, we have to serialize the keys in a particular order.

Reply:
I think if we let Portal owners mint a DOI for a resource, then we must have a contract that the primary key for that resource can no longer change.